### ISSUE TITLE
yum performance issue installing packages

### ISSUE THREAD

**Issue by furlongm** [saltstack/salt#44305]
### Description of Issue/Question

There seems to be a yum performance issue when there is one or more packages to be installed. This causes the whole run to take (in my case) 8 minutes. The equivalent puppet and chef runs take 2 minutes to run.

This is because salt iterates over ['yum', '--quiet', 'repository-packages', 'repo_name', 'list', '--showduplicates'] for every repo, and for every package. This takes > 5 seconds per yum invocation, per package. This seems to happen for _every_ package even if there is only one package to be installed (i.e. for packages that are already installed but are part of the salt run).

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)

### Steps to Reproduce Issue

Compare the time taken to run the `yum repository-packages <package>` command (in salt and yum) vs running the `yum list available <package>` command in yum:

```
# time salt-call --local pkg.list_repo_pkgs bash -l info
[INFO    ] Executing command ['yum', '--version'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'rhel-server-rhscl-7-rpms', 'list', '--showduplicates', 'bash'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'rhel-7-server-optional-rpms', 'list', '--showduplicates', 'bash'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'rhel-7-server-rh-common-rpms', 'list', '--showduplicates', 'bash'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'rhel-7-server-extras-rpms', 'list', '--showduplicates', 'bash'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'rhel-7-server-dotnet-rpms', 'list', '--showduplicates', 'bash'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'epel-7', 'list', '--showduplicates', 'bash'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'rhel-7-server-rpms', 'list', '--showduplicates', 'bash'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'saltstack', 'list', '--showduplicates', 'bash'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'rhel-7-server-supplementary-rpms', 'list', '--showduplicates', 'bash'] in directory '/root'
[INFO    ] Executing command ['yum', '--quiet', 'repository-packages', 'rhel-7-server-thirdparty-oracle-java-rpms', 'list', '--showduplicates', 'bash'] in directory '/root'
local:
    ----------
    bash:
        - 4.2.46-29.el7_4
        - 4.2.46-28.el7
        - 4.2.46-21.el7_3
        - 4.2.46-20.el7_2
        - 4.2.46-19.el7
        - 4.2.46-12.el7
        - 4.2.45-5.el7_0.4
        - 4.2.45-5.el7_0.2
        - 4.2.45-5.el7

real    0m5.150s
user    0m4.163s
sys     0m0.982s
```

Running the same commands via command line takes around the same amount of time (per package):

```
# time for i in $(yum --quiet repolist enabled  | grep -v "^repo id" | cut -d " " -f 1 | cut -d / -f 1); do yum --quiet repository-packages ${i} list --showduplicates bash 2>/dev/null; done

Installed Packages
bash.x86_64                                                                4.2.46-29.el7_4                                                                 @rhel-7-server-rpms
Available Packages
bash.x86_64                                                                4.2.45-5.el7                                                                    rhel-7-server-rpms 
bash.x86_64                                                                4.2.45-5.el7                                                                    rhel-7-server-rpms 
bash.x86_64                                                                4.2.45-5.el7                                                                    rhel-7-server-rpms 
bash.x86_64                                                                4.2.45-5.el7_0.2                                                                rhel-7-server-rpms 
bash.x86_64                                                                4.2.45-5.el7_0.2                                                                rhel-7-server-rpms 
bash.x86_64                                                                4.2.45-5.el7_0.2                                                                rhel-7-server-rpms 
bash.x86_64                                                                4.2.45-5.el7_0.4                                                                rhel-7-server-rpms 
bash.x86_64                                                                4.2.45-5.el7_0.4                                                                rhel-7-server-rpms 
bash.x86_64                                                                4.2.45-5.el7_0.4                                                                rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-12.el7                                                                   rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-12.el7                                                                   rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-12.el7                                                                   rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-19.el7                                                                   rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-19.el7                                                                   rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-19.el7                                                                   rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-20.el7_2                                                                 rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-20.el7_2                                                                 rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-20.el7_2                                                                 rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-21.el7_3                                                                 rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-21.el7_3                                                                 rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-21.el7_3                                                                 rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-28.el7                                                                   rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-28.el7                                                                   rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-28.el7                                                                   rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-29.el7_4                                                                 rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-29.el7_4                                                                 rhel-7-server-rpms 
bash.x86_64                                                                4.2.46-29.el7_4                                                                 rhel-7-server-rpms 

real    0m4.298s
user    0m3.473s
sys     0m0.826s
```

Compare that to just running the `list available package` command:

```# time yum --quiet --showduplicates list available bash | uniq
Available Packages
bash.x86_64                 4.2.45-5.el7                      rhel-7-server-rpms
bash.x86_64                 4.2.45-5.el7_0.2                  rhel-7-server-rpms
bash.x86_64                 4.2.45-5.el7_0.4                  rhel-7-server-rpms
bash.x86_64                 4.2.46-12.el7                     rhel-7-server-rpms
bash.x86_64                 4.2.46-19.el7                     rhel-7-server-rpms
bash.x86_64                 4.2.46-20.el7_2                   rhel-7-server-rpms
bash.x86_64                 4.2.46-21.el7_3                   rhel-7-server-rpms
bash.x86_64                 4.2.46-28.el7                     rhel-7-server-rpms
bash.x86_64                 4.2.46-29.el7_4                   rhel-7-server-rpms

real    0m0.526s
user    0m0.396s
sys     0m0.130s
```

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)

```
# salt-call --versions-report
Salt Version:
           Salt: 2017.7.2
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, May  3 2017, 07:55:04)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: redhat 7.4 Maipo
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-693.5.2.el7.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 7.4 Maipo
```

**Comment by johnnyrun** [comment]
same issue here.
Seems called on every package  here:

https://github.com/saltstack/salt/blob/2017.7.2/salt/modules/yumpkg.py#L1284

develop branch is very different...

**Comment by Ch3LL** [comment]
You say develop branch is different. When using develop do you get faster times as well?

**Comment by junster1** [comment]
We are seeing the same issues. It looks like you are gathering the byrepo data even if you are are not using the data.

https://github.com/saltstack/salt/blob/2017.7.2/salt/modules/yumpkg.py#L859

**Comment by junster1** [comment]
This seems to have been fixed with PR #43396

**Comment by terminalmage** [comment]
Correction, the PR was actually https://github.com/saltstack/salt/pull/43687. The fix went into the 2017.7.3 release.
